bin/propolis-server/src/lib/server.rs

```diff
@@ -458,6 +458,26 @@ async fn instance_ensure(
         }));
     }
 
+    let log = server_context.log.clone();
+    let mut state_watcher = vm.state_watcher().clone();
+    let services = Arc::clone(&server_context.services);
+    tokio::task::spawn(async move {
+        loop {
+            if let Err(_) = state_watcher.changed().await {
+                // The sender has been dropped (i.e. the VM has been destroyed).
+                // Nothing more for us to do so just bail.
+                break;
+            }
+
+            let state = state_watcher.borrow().state;
+            if state == api::InstanceState::Stopping {
+                // The VM has started Stopping, release the per-instance service providers
```

Comment:
Are we guaranteed to observe this transition? My understanding of the tokio `watch` is that you could have something like this happen:

- this task calls `state_watcher.changed` and awaits
- VM controller writes `Stopping` to the channel, satisfying the call to `changed`
- before the task gets to run again, the VM controller finishes halting the VM and writes `Stopped` to the channel
- this task calls `state_watcher.borrow()` and gets `Stopped`, so doesn't reach line 475

Assuming I'm not misreading the `watch::Receiver` docs, would it make sense to create a separate channel for "the guest shut itself down" messages?

Reply:
Nice catch! I swapped it to use a one-shot channel that gets explicitly used once the state worker has finished.